### PR TITLE
Add legend title font properties

### PR DESCRIPTION
**Pull request by smartlixx**
## PR Summary
Add a keyword ```title_fontproperties``` to `matplotlib.pyplot.legend`, so that we can prescribe font properties for legend's title. It can't be used together with the `title_fontsize`.  This closes issue #19259.

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

Now we can use the following codes to specify legend title font properties, and `title_fontsize` will override the `size` in `title_prop`. 

```python
import matplotlib.pyplot as plt
import matplotlib as mpl

myfont1 = mpl.font_manager.FontProperties(fname='/System/Library/Fonts/Supplemental/Arial Italic.ttf')
myfont2 = mpl.font_manager.FontProperties(fname='/System/Library/Fonts/Supplemental/Arial Bold.ttf',size=20)
myfont3 = mpl.font_manager.FontProperties(fname='/System/Library/Fonts/Supplemental/Arial Bold.ttf')

fig, [[ax1, ax2], [ax3, ax4]] = plt.subplots(2, 2, figsize=(8, 6))

ax1.scatter(1, 1, 40, label='point')
ax1.legend(title='location')

ax2.scatter(1, 1, 40, label='point')
ax2.legend(title='location', prop=myfont1, title_fontproperties=myfont2)

ax3.scatter(1, 1, 40, label='point')
ax3.legend(title='location', prop=myfont1, title_fontsize=20)

ax4.scatter(1, 1, 40, label='point')
ax4.legend(title='location', prop=myfont1, title_fontproperties={'weight':'bold', 'size':20})

plt.show()
``` 

The output is 

test_legend_title_prop

### CODE REVIEW

**Comment by jklymak** [comment]
Thanks a lot, this seems good (without looking at it too hard).  It would be nice if the new functionality were in our automated tests,  it probably needs a "what's new" and maybe an example in the docs?  Perhaps just extend an example that already sets the fontsize?

File: lib/matplotlib/legend.py
Comment (jklymak) [inline lib/matplotlib/legend.py:525]: Thanks for the tests.  It would be good to make sure all the patch are tested a) so we understand the behaviour and b) so no one removes it later...

File: doc/users/next_whats_new/legend_title_prop_kwarg.rst
Comment (jklymak) [inline doc/users/next_whats_new/legend_title_prop_kwarg.rst:9]: rather than abstract, can you just put an example of the dict in here?

Comment (smartlixx) [reply doc/users/next_whats_new/legend_title_prop_kwarg.rst:9]: Thanks. I will add an example of the dict.

**Comment by smartlixx** [comment]
I don't know how to address the failures of docs building. Could anyone help, please?

File: doc/users/next_whats_new/legend_title_prop_kwarg.rst
Comment (aitikgupta) [inline doc/users/next_whats_new/legend_title_prop_kwarg.rst:7]:
```suggestion
as set by the :rc:`legend.title_fontsize`. Through this new kwarg, one
```

File: doc/users/next_whats_new/legend_title_prop_kwarg.rst
Comment (aitikgupta) [inline doc/users/next_whats_new/legend_title_prop_kwarg.rst:14]: I think it'll be better to put this in the `plot` directive?

```suggestion

.. plot::
    fig, ax = plt.subplots(figsize=(4, 3))
    ax.plot(range(10)
    ax.legend(title='Points', title_prop={'family': 'serif'}, title_fontsize=22) 
```

**Review by aitikgupta — COMMENTED** [review]
Docs fail with following warnings:
```console
/home/circleci/project/doc/users/next_whats_new/legend_title_prop_kwarg.rst:7: WARNING: Unexpected indentation.
/home/circleci/project/doc/users/next_whats_new/legend_title_prop_kwarg.rst:8: WARNING: Block quote ends without a blank line; unexpected unindent.
```

**Comment by smartlixx** [comment]
> Docs fail with following warnings:
> 
> ```
> /home/circleci/project/doc/users/next_whats_new/legend_title_prop_kwarg.rst:7: WARNING: Unexpected indentation.
> /home/circleci/project/doc/users/next_whats_new/legend_title_prop_kwarg.rst:8: WARNING: Block quote ends without a blank line; unexpected unindent.
> ```

Many thanks. I have accepted the suggested changes.

File: doc/users/next_whats_new/legend_title_prop_kwarg.rst
Comment (aitikgupta) [inline doc/users/next_whats_new/legend_title_prop_kwarg.rst:12]:
```
/home/circleci/project/doc/users/next_whats_new/legend_title_prop_kwarg.rst:11: WARNING: Error in "plot" directive:
maximum 2 argument(s) allowed, 10 supplied.

.. plot::
    fig, ax = plt.subplots(figsize=(4, 3))
    ax.plot(range(10)
    ax.legend(title='Points', title_prop={'family': 'serif'}, title_fontsize=22)
```
You might wanna add a newline after `.. plot::`

```suggestion
.. plot::

    fig, ax = plt.subplots(figsize=(4, 3))
```

File: lib/matplotlib/legend.py
Comment (smartlixx) [reply lib/matplotlib/legend.py:525]: I added one more test for Line 528, but can't get the test for Line 524 right. I checked `mpl.rcParams['legend.title_fontsize']` but find it is set to `None`. So the test `assert leg.get_title().get_fontsize() == mpl.rcParams['legend.title_fontsize']` will always fail.

File: lib/matplotlib/legend.py
Comment (jklymak) [inline lib/matplotlib/legend.py:530]: This logic seems convoluted to me and checks the same thing a few different times.  Can it be simplified?  

```suggestion
        self.title_prop = FontProperties._from_any(title_prop)
        if isinstance(title_prop, dict) and "size" not in title_prop:
            if title_fontsize is None:
                title_fontsize = mpl.rcParams["legend.title_fontsize"]
            self.title_prop.set_size(title_fontsize)
        else:
            if title_fontsize is not None:
                self.title_prop.set_size(title_fontsize)

        self.set_title(title, prop=self.title_prop)
```

Comment (smartlixx) [reply lib/matplotlib/legend.py]: Seems that _from_any can take `None` as its input. Now it looks simpler.

File: lib/matplotlib/legend.py
Comment (jklymak) [reply lib/matplotlib/legend.py:522]: So now line 518 needs a check.  But I think this is very close!  Thanks for your patience...

Comment (smartlixx) [reply lib/matplotlib/legend.py:522]: I just added one more test for L518. In this case when rcParams['legend.title_fontsize'] is `None`, the title fontsize will be defaulted to rcParams['font.size'] (when a FontProperties object is created with default args).

**Review by jklymak — APPROVED** [review]
Thanks for sticking with it.  Looks good to me, but we need a second review.

File: lib/matplotlib/legend.py
Comment (QuLogic) [inline lib/matplotlib/legend.py:330]: This must be inserted at the end to not break the API, until we make these keyword-only arguments.

File: lib/matplotlib/tests/test_legend.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_legend.py:547]: Test name could be updated.

**Review by QuLogic — COMMENTED** [review]
I suppose it might make sense to deprecate `title_fontsize` at this point? @timhoffm

File: lib/matplotlib/tests/test_legend.py
Comment (smartlixx) [reply lib/matplotlib/tests/test_legend.py]: All suggested changes are done.

File: lib/matplotlib/legend.py
Comment (tacaswell) [inline lib/matplotlib/legend.py:515]: Do we need to hold onto this and expose it as public API?  Previously we did not hold onto it and it does not look like that the user can update this after the fact and have it be respected.

Comment (tacaswell) [inline lib/matplotlib/legend.py:515]:
```suggestion
        title_prop = FontProperties._from_any(title_prop)
```

File: lib/matplotlib/legend.py
Comment (tacaswell) [inline lib/matplotlib/legend.py:519]:
```suggestion
            title_prop.set_size(title_fontsize)
```

File: lib/matplotlib/legend.py
Comment (tacaswell) [inline lib/matplotlib/legend.py:524]:
```suggestion
        set_title(title, prop=title_prop)
```

File: lib/matplotlib/legend.py
Comment (tacaswell) [inline lib/matplotlib/legend.py:522]:
```suggestion
                title_prop.set_size(title_fontsize)
```

**Review by tacaswell — DISMISSED** [review]
I do not think we should add `self.title_prop` as new public API on the `Legend`.

File: lib/matplotlib/legend.py
Comment (smartlixx) [reply lib/matplotlib/legend.py]: For the sake of L516 (it checks the original value of 'title_prop' argument, I add a new local variable 'title_prop_fp' here.

File: lib/matplotlib/legend.py
Comment (timhoffm) [inline lib/matplotlib/legend.py:240]:
```suggestion
title_fontproperties : None or `matplotlib.font_manager.FontProperties` or dict
```
The naming is consistent with the parameter `prop`. Unfortunately, this is inconsistent with the rest of the library. In all other places we use `fontproperties`, not `prop`.

IMO `title_fontproperties` is to be preferred because it's clearer. It's a bit lengthy, but it's only used for fine-tuning and the passed value will likely be verbose as well. So we should favor clarity here.

Comment (jklymak) [reply lib/matplotlib/legend.py:240]: Yeah, I agree - good catch...

Comment (smartlixx) [reply lib/matplotlib/legend.py]: Thanks for the suggestion. I have modified this parameter's name to title_fontproperties.

File: lib/matplotlib/legend.py
Comment (timhoffm) [inline lib/matplotlib/legend.py:523]: I spent several minutes trying to figure what the precedence rules are and if they are reasonable.

I propose to forbid using both `title_fontsize` and `title_fontproperties` together. The combinations are awkward anyway: Specifying size in both is unclear; specifying size explicitly but e.g. family via `title_fontproperties` is clumsy - one should specify both via `title_fontproperties` in such a case.

The behavior should also be mentioned in the docstring.

File: doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst
Comment (timhoffm) [inline doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst:2]:
```suggestion
Font properties of legend title are configurable
------------------------------------------------
```
Better have a semantic title than a technical one

File: doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst
Comment (timhoffm) [inline doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst:9]:
```suggestion
title's font properties set via the ``title_fontproperties`` keyword
argument, for example:
```
This description should only give the user an idea about the new feature. Details like supported formats and defaults are not relevant here. They can be looked up easily using the links above.

File: doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst
Comment (timhoffm) [inline doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst:15]:
```suggestion
    ax.legend(title='Points', title_fontproperties={'family': 'serif', 'fontsize': 22}) 
```
As discussed below, I wouldn't allow `title_fontproperties` and `title_fontsize` simultaneously.

File: lib/matplotlib/legend.py
Comment (smartlixx) [reply lib/matplotlib/legend.py:523]: Or should we just deprecate `title_fontsize`, as suggested by @QuLogic above?

Comment (QuLogic) [reply lib/matplotlib/legend.py:523]: @timhoffm?

Even if we deprecate though, we'd probably need to decide on precedence in the interim.

Comment (timhoffm) [reply lib/matplotlib/legend.py:523]: I wouldn't not to deprecate `title_fontsize` for now. It's been in since 3.0 and forcing all users to switch is not justified. Also I'm slightly hesitant because `title_fontproperties={'size': 12}` is much more to write compared to `title_fontsize=12`.

If we forbid using both ways simultaneously, we don't have to decide on precedence.

**Comment by timhoffm** [comment]
@smartlixx You've requested a review, however there are still unanswered comments from the previous review. Please address all questions before requesting a review.

**Review by timhoffm — COMMENTED** [review]
@smartlixx You've requested a review, however there are still unanswered comments from the previous review. Please address all questions before requesting a review.

**Comment by smartlixx** [comment]
> @smartlixx You've requested a review, however there are still unanswered comments from the previous review. Please address all questions before requesting a review.

Thanks. I was waiting for your confirmation whether we should deprecate `title_fontsize`. Now that it is confirmed, I have made changes accordingly.

File: lib/matplotlib/legend.py
Comment (smartlixx) [reply lib/matplotlib/legend.py]: OK, thanks for your confirmation.

File: lib/matplotlib/legend.py
Comment (timhoffm) [inline lib/matplotlib/legend.py:246]:
```suggestion
    The font size of the legend's title.
    Note: This cannot be combined with *title_fontproperties*. If you want
    to set the fontsize alongside other font properties, use the *size*
    parameter in font properties.
```

**Review by timhoffm — APPROVED** [review]
Modulo one doc clarification.

**Comment by smartlixx** [comment]
@tacaswell can you please have a look? I still see the 'changes requested' label after I made those changes.

PS: Should I squash the commits to reduce the commit history?

File: doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst
Comment (QuLogic) [inline doc/users/next_whats_new/legend_title_fontproperties_kwarg.rst:4]:
```suggestion
Title's font properties can be set via the *title_fontproperties* keyword
```

File: lib/matplotlib/legend.py
Comment (QuLogic) [inline lib/matplotlib/legend.py:242]: Which rcParams?

File: lib/matplotlib/legend.py
Comment (QuLogic) [inline lib/matplotlib/legend.py:531]: This is not right, as if you pass in a `FontProperties`, the size will be overridden.

**Comment by QuLogic** [comment]
You will need to rebase to fix the doc build.

File: lib/matplotlib/legend.py
Comment (smartlixx) [reply lib/matplotlib/legend.py:531]: Thanks for pointing out this. I have made the necessary changes. For `FontProperties`, there seems to have no way to determine if a `size` parameter is prescribed. So for `FontProperties`, no change of title fontsize will be performed.

**Comment by smartlixx** [comment]
After rebase, the doc build still fails, at different places. And the Azure jobs failure is not related to this PR.

File: lib/matplotlib/legend.py
Comment (timhoffm) [inline lib/matplotlib/legend.py:242]:
```suggestion
    *title_fontsize* argument will be used if present; if *title_fontsize* is
```

File: lib/matplotlib/legend.py
Comment (timhoffm) [inline lib/matplotlib/legend.py:244]:
```suggestion
    :rc:`matplotlib.rcParams["legend.title_fontsize"]` will be used.
```

**Review by timhoffm — COMMENTED** [review]
Should fix the docs.

File: lib/matplotlib/tests/test_legend.py
Comment (QuLogic) [inline lib/matplotlib/tests/test_legend.py:555]: Only the `legend` raises here:
```suggestion
    plt.plot(range(10))
    with pytest.raises(ValueError):
        plt.legend(title='Aardvark', title_fontsize=22,
                   title_fontproperties={'family': 'serif', 'size': 22})
```

**Review by QuLogic — COMMENTED** [review]
Some small things could be shortened.

**Pull request merged**
